### PR TITLE
8300916: Re-examine the initialization of JNU Charset in StaticProperty

### PR DESCRIPTION
**Pull request by naotoj**
This issue was found during the review of this PR: https://github.com/openjdk/jdk/pull/12132 where `Charset` class was loaded/initialized at the phase 1 of the startup process. Since `Charset` depends on `StaticProperty`, loading of `Charset` class should be delayed. I basically moved cache for `jnuCharset` into the actual calling locations `ProcessImpl` and `ProcessEnvironment` for unix platforms so that initPhase1() won't initialize `Charset` class.
Unrelated, but I replaced `Locale.ENGLISH` with `Locale.ROOT` in the argument of `toLowerCase()`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8301114](https://bugs.openjdk.org/browse/JDK-8301114) to be approved

### Issues
 * [JDK-8300916](https://bugs.openjdk.org/browse/JDK-8300916): Re-examine the initialization of JNU Charset in StaticProperty
 * [JDK-8301114](https://bugs.openjdk.org/browse/JDK-8301114): Re-examine the initialization of JNU Charset in StaticProperty (**CSR**)

### Reviewers
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to [3f9165bb](https://git.openjdk.org/jdk/pull/12171/files/3f9165bb739305d7cb54024d415206b110d5a4ca)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [3f9165bb](https://git.openjdk.org/jdk/pull/12171/files/3f9165bb739305d7cb54024d415206b110d5a4ca)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12171/head:pull/12171` \
`$ git checkout pull/12171`

Update a local copy of the PR: \
`$ git checkout pull/12171` \
`$ git pull https://git.openjdk.org/jdk pull/12171/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12171`

View PR using the GUI difftool: \
`$ git pr show -t 12171`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12171.diff">https://git.openjdk.org/jdk/pull/12171.diff</a>

</details>

### CODE REVIEW

File: src/java.base/unix/classes/java/lang/ProcessImpl.java [hunk 1]
```diff
@@ -69,6 +69,10 @@ final class ProcessImpl extends Process {
     // Linux platforms support a normal (non-forcible) kill signal.
     static final boolean SUPPORTS_NORMAL_TERMINATION = true;
 
+    // Cache for JNU Charset. The encoding name is guaranteed
+    // to be supported in this environment.
+    static final Charset JNU_CHARSET = Charset.forName(StaticProperty.jnuEncoding());
```
Comment (jaikiran) [inline src/java.base/unix/classes/java/lang/ProcessImpl.java:74]: Before the change in this PR, the `jnuCharset` used to fallback to `Charset.defaultCharset()`, if the `sun.jnu.encoding` value wasn't a legal charset or couldn't be loaded. 
On this line, we no longer do that. The comment, above this line, indicates that this is intentional. Looking at the code in `java.lang.System#initPhase1()`, the `sun.jnu.encoding` property value is updated to `UTF-8` if the previous value isn't a supported charset. That is done just before the `StaticProperty` class is referenced and loaded. Effectively, as noted in this comment, the `sun.jnu.encoding` would just have the right usable/supported value stored in `StaticProperty.jnuEncoding()` by the time we reach here. So this change looks fine to me.

**Comment by jaikiran** [comment]
Hello Naoto,

> This issue was found during the review of this PR: https://github.com/openjdk/jdk/pull/12132 where Charset class was loaded/initialized at the phase 1 of the startup process. Since Charset depends on StaticProperty

From the comment in that PR:

> That said, the change does highlight an issue in StaticProperty. where it calls Charset.defaultCharset(), which in turn will StaticProperty.FILE_ENCODING before that class is fully initialized.

Looking at the code in `Charset.defaultCharset()` https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/nio/charset/Charset.java#L651, I don't see it using `StaticProperty` class. Instead, it just queries the `System.getProperty()`. I don't see any other methods in `Charset` using `StaticProperty` either.

**Comment by AlanBateman** [comment]
The change to StaticProperty to avoid calling out to Charset.defaultCharset from the initializer is good. However, the other part to that is the scenario in PR 12132 where the default Charset was accidentally located via the provider mechanism in JDK 9-17. If I read the changes correctly, that fragile scenario will come back. We have a couple of ways to avoid that, one being to ensure that defaultCharset is called before the boot layer is set. A simpler, and more reliable, would be to change Charset.defaultCharset to use StandardCharsets.charsetForName with the value of "file.encoding", and avoid the provider lookup completely.

**Comment by AlanBateman** [comment]
> `Charset.defaultCharset()` now uses `standardProvider.charsetForName(<file.encoding>)` charset.

I think this is the right thing to do. It can also be changed to use StaticProperty.fileEncoding() and maybe the field can be changed to be a `@Stable` field. 

It might be that we will need to create a CSR and Release Note for this change.  The scenario is PR 12132 is unfortunate but does not show that some deployments may have been relying on the this from JDK 9 to JDK 17. With the change here, we are doubling now on ensuring that the default charset is loaded from java.base.

**Comment by naotoj** [comment]
Good point. Will make the change and create a CSR/relnote.

**Comment by mlchung** [comment]
> A simpler, and more reliable, would be to change Charset.defaultCharset to use standardProvider.charsetForName with the value of "file.encoding", and avoid the provider lookup completely.

This is a good observation.   The change looks good to me.

File: src/java.base/share/classes/java/nio/charset/Charset.java [hunk 2]
```diff
@@ -628,6 +629,7 @@ public SortedMap<String,Charset> run() {
             });
     }
 
+    @Stable
     private static volatile Charset defaultCharset;
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/nio/charset/Charset.java:633]: If the field is stable then does the volatile go away?

Comment (naotoj) [reply src/java.base/share/classes/java/nio/charset/Charset.java:633]: Right. It should.

File: src/java.base/share/classes/java/nio/charset/Charset.java [hunk 3]
```diff
@@ -648,12 +649,19 @@ public SortedMap<String,Charset> run() {
...
-                String csn = GetPropertyAction
-                        .privilegedGetProperty("file.encoding");
-                Charset cs = lookup(csn);
+                // do not look for providers other than the standard one
+                Charset cs;
+                if (System.getSecurityManager() == null) {
+                    cs = standardProvider.charsetForName(StaticProperty.fileEncoding());
+                } else {
+                    PrivilegedAction<Charset> pa =
+                        () -> standardProvider.charsetForName(StaticProperty.fileEncoding());
```
Comment (mlchung) [inline src/java.base/share/classes/java/nio/charset/Charset.java:662]: what operation does this do security permission check?  I don't think any.

Comment (naotoj) [reply src/java.base/share/classes/java/nio/charset/Charset.java:662]: Right. I was under the impression that the call should have been made with privileged action (as in the comments for those access methods in `StaticProperty`, but the property value itself is not given to untrusted code. Will remove the permission check.

**Review by mlchung — APPROVED** [review]

**Comment by takiguc** [comment]
Hello @AlanBateman 
You said

> The change to StaticProperty to avoid calling out to Charset.defaultCharset from the initializer is good. However, the other part to that is the scenario in PR 12132 where the default Charset was accidentally located via the provider mechanism in JDK 9-17. If I read the changes correctly, that fragile scenario will come back. We have a couple of ways to avoid that, one being to ensure that defaultCharset is called before the boot layer is set. A simpler, and more reliable, would be to change Charset.defaultCharset to use standardProvider.charsetForName with the value of "file.encoding", and avoid the provider lookup completely.

Could you explain about fragile scenario ?

**Comment by AlanBateman** [comment]
> Could you explain about fragile scenario ?

Charset.defaultCharset does the lookup on the first usage. Since JDK 18, that first usage is in initPhase1 so it will always find the default charset in java.base. In JDK 9-17, the first usage may have been in initPhase2 (during module system initialization), in which case it would locate the default charset in java.base, the first usage may have been after the VM is fully initialized, in which case it would be a service provider lookup and maybe find it in jdk.charsets or on the class path.

I've added a comment to the [CSR](https://bugs.openjdk.org/browse/JDK-8301114) with more context and to support the proposal to double down on requiring the default charset to be in java.base.

File: src/java.base/share/classes/java/nio/charset/Charset.java [hunk 4]
```diff
@@ -628,7 +628,8 @@ public SortedMap<String,Charset> run() {
             });
     }
 
-    private static volatile Charset defaultCharset;
+    @Stable
+    private static Charset defaultCharset;
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/nio/charset/Charset.java:632]: Style wise, I think the annotation is added between the private and static modifiers in other places.

**Review by AlanBateman — APPROVED** [review]
Thanks for the updates/iterations, I think you've got this to a good place.

One thing to think about is having System.initPhase3 read file.encoding and if not UTF-8, it could call Charset.defaultCharset and if not the expected value then it could emit a warning like is done for a bad value of java.io.tmpdir.

One thing is whether to add a regression test to ensure that the default charset is UTF-8 when run with -Dfile.encoding=XXX and XXX is in the service provider module.

**Comment by naotoj** [comment]
> Thanks for the updates/iterations, I think you've got this to a good place.
> 
> One thing to think about is having System.initPhase3 read file.encoding and if not UTF-8, it could call Charset.defaultCharset and if not the expected value then it could emit a warning like is done for a bad value of java.io.tmpdir.
> 
> One thing is whether to add a regression test to ensure that the default charset is UTF-8 when run with -Dfile.encoding=XXX and XXX is in the service provider module.

Filed: https://bugs.openjdk.org/browse/JDK-8301199

File: src/java.base/share/classes/java/nio/charset/Charset.java (hunk 4 above)
Comment (naotoj) [reply src/java.base/share/classes/java/nio/charset/Charset.java]: Fixed.

**Pull request closed without merging**